File: fraiseql-python/src/fraiseql/federation.py
"""Federation support for Apollo Federation v2 in FraiseQL.

This module provides decorators for defining federated GraphQL schemas:
- @key: Define federation keys for entity resolution
- @extends: Extend types from other subgraphs
- @external: Mark fields as owned by other subgraphs
- @requires: Declare field dependencies
- @provides: Mark fields that provide data for other subgraphs

Example:
    ```python
    from fraiseql import type as fraiseql_type
    from fraiseql.federation import key, extends, external, requires, provides

    # Authoritative User type in this subgraph
    @fraiseql_type
    @key("id")
    class User:
        id: str
        email: str

    # Extended User type in another subgraph
    @fraiseql_type
    @extends
    @key("id")
    class User:
        id: str = external()
        email: str = external()
        orders: list[Order]
    ```
"""

from __future__ import annotations

import inspect
from typing import TYPE_CHECKING, Any, TypeVar

from fraiseql.errors import FederationValidationError

if TYPE_CHECKING:
    from collections.abc import Callable

T = TypeVar("T")

# Federation metadata keys
_KEYS = "keys"
_EXTEND = "extend"
_EXTERNAL_FIELDS = "external_fields"
_REQUIRES = "requires"
_PROVIDES_DATA = "provides_data"


def _init_federation_metadata() -> dict[str, Any]:
    """Initialize an empty federation metadata dictionary."""
    return {
        _KEYS: [],
        _EXTEND: False,
        _EXTERNAL_FIELDS: [],
        _REQUIRES: {},
        _PROVIDES_DATA: [],
    }


def _get_or_init_federation_metadata(cls: type[T]) -> dict[str, Any]:
    """Get or initialize federation metadata on a class."""
    if not hasattr(cls, "__fraiseql_federation__"):
        cls.__fraiseql_federation__ = _init_federation_metadata()
    return cls.__fraiseql_federation__


def _check_type_decorator_applied(cls: type[T]) -> bool:
    """Check if @fraiseql.type or @type decorator is in the decorator stack.

    This is a heuristic check using source inspection.
    Returns True if type decorator appears to be present.
    """
    try:
        source = inspect.getsource(cls)
        # Check if fraiseql_type or just 'type' decorator appears before class definition
        lines = source.split("\n")
        return any("@" in line and "type" in line.lower() for line in lines)
    except (OSError, TypeError):
        # Can't get source (e.g., in tests or REPL), assume it's OK
        return True


def _extract_key_fields(metadata: dict[str, Any]) -> set[str]:
    """Extract all field names used in keys from federation metadata."""
    key_fields: set[str] = set()
    for key_def in metadata.get(_KEYS, []):
        key_fields.update(key_def.get("fields", []))
    return key_fields


def _validate_field_exists(field_name: str, annotations: dict[str, Any], class_name: str) -> None:
    """Validate that a field name exists in class annotations."""
    if field_name not in annotations:
        raise FederationValidationError(f"Field '{field_name}' not found in {class_name}")


def _collect_field_markers(
    cls: type[T],
) -> tuple[set[str], dict[str, FieldDefault]]:
    """Collect all field markers and their metadata from a class.

    Returns:
        Tuple of (field_names_with_markers, marker_by_field)
    """
    annotations = getattr(cls, "__annotations__", {})
    field_markers: dict[str, FieldDefault] = {}
    marked_fields: set[str] = set()

    for field_name in annotations:
        field_default = getattr(cls, field_name, None)
        if isinstance(field_default, FieldDefault):
            field_markers[field_name] = field_default
            marked_fields.add(field_name)

    return marked_fields, field_markers


def _process_field_markers(
    metadata: dict[str, Any],
    field_markers: dict[str, FieldDefault],
    annotations: dict[str, Any],
    class_name: str,
    is_extended: bool = False,
) -> None:
    """Process field markers and update federation metadata.

    Handles @requires, @provides, and @external markers.
    """
    for field_name, marker in field_markers.items():
        # Handle @requires()
        if marker.requires:
            _validate_field_exists(marker.requires, annotations, class_name)
            metadata[_REQUIRES][field_name] = marker.requires

        # Handle @provides()
        if marker.provides:
            metadata[_PROVIDES_DATA].extend(marker.provides)

        # Handle @external() - only relevant for extended types
        if is_extended and marker.external:
            _validate_field_exists(field_name, annotations, class_name)
            metadata[_EXTERNAL_FIELDS].append(field_name)


class FieldDefault:
    """Marker for federation field metadata (external, requires, provides).

    This class is used as a default value in field annotations to store
    federation metadata on the field.

    Attributes:
        external: Whether field is owned by another subgraph.
        requires: Field name that must be resolved first.
        provides: List of external fields this field provides data for.
    """

    def __init__(
        self,
        external: bool = False,
        requires: str | None = None,
        provides: list[str] | None = None,
    ) -> None:
        self.external = external
        self.requires = requires
        self.provides = provides or []


def external() -> FieldDefault:
    """Mark a field as external (owned by another subgraph).

    Use this in extended types to mark which fields are owned by the
    authoritative subgraph.

    Returns:
        FieldDefault marked as external.

    Example:
        ```python
        @fraiseql_type
        @extends
        @key("id")
        class User:
            id: str = external()
            email: str = external()
            orders: list[Order]  # New field in this subgraph
        ```

    Raises:
        ValueError: If used on non-extended types or if field doesn't exist.
    """
    return FieldDefault(external=True)


def requires(field_name: str | list[str]) -> FieldDefault:
    """Mark a field as requiring another field to be resolved first.

    This declares that a field needs data from another field (in the same
    type or from federation) to compute its value.

    Args:
        field_name: Name of the field(s) that must be resolved first.

    Returns:
        FieldDefault with requires dependency.

    Example:
        ```python
        @fraiseql_type
        @extends
        @key("id")
        class User:
            id: str = external()
            email: str = external()
            profile: UserProfile = requires("email")  # Needs email to resolve
        ```

    Raises:
        ValueError: If the referenced field doesn't exist.
    """
    # Normalize field_name to string
    if isinstance(field_name, list):
        if len(field_name) != 1:
            raise FederationValidationError("@requires supports only single field dependency")
        field_name = field_name[0]
    return FieldDefault(requires=field_name)


def provides(*targets: str) -> FieldDefault:
    """Mark a field as providing data for external subgraph fields.

    This declares that this field's data can be used to resolve fields
    in other subgraphs.

    Args:
        *targets: List of "Type.field" references this field provides data for.

    Returns:
        FieldDefault with provides targets.

    Example:
        ```python
        @fraiseql_type
        @key("id")
        class User:
            id: str
            email: str = provides("Order.owner_email", "Invoice.owner_email")
        ```
    """
    return FieldDefault(provides=list(targets))


def key(field_names: str | list[str]) -> Callable[[type[T]], type[T]]:
    """Mark a type with a federation key for entity resolution.

    Federation keys are used to uniquely identify entities and resolve them
    across subgraphs. Multiple @key decorators define composite keys.

    Args:
        field_names: Field name or list of field names that form the key.

    Returns:
        Decorator function that stores key metadata on the class.

    Example:
        ```python
        @fraiseql_type
        @key("id")
        class User:
            id: str
            email: str

        @fraiseql_type
        @key("tenant_id")
        @key("id")
        class Account:
            tenant_id: str
            id: str
            name: str
        ```

    Raises:
        ValueError: If key field doesn't exist on the type.
    """
    # Normalize field_names to list
    if isinstance(field_names, str):
        field_names = [field_names]

    def decorator(cls: type[T]) -> type[T]:
        # Validate that @type decorator was applied or will be applied
        if not _check_type_decorator_applied(cls):
            raise TypeError(f"@key requires @type decorator to be applied to {cls.__name__}")

        # Get or initialize federation metadata
        metadata = _get_or_init_federation_metadata(cls)
        annotations = getattr(cls, "__annotations__", {})

        # Validate all key fields exist
        for field_name in field_names:
            _validate_field_exists(field_name, annotations, cls.__name__)

        # Check for duplicate keys
        new_key = {"fields": field_names}
        if new_key in metadata[_KEYS]:
            raise FederationValidationError(f"Duplicate key field in {cls.__name__}")

        # Add key to federation metadata
        metadata[_KEYS].append(new_key)

        # Collect and process field markers on non-extended types
        # Note: Extended types are processed in @extends decorator
        if not metadata[_EXTEND]:
            _, field_markers = _collect_field_markers(cls)
            _process_field_markers(metadata, field_markers, annotations, cls.__name__)

        return cls

    return decorator


def extends(cls: type[T] | None = None) -> type[T] | Callable[[type[T]], type[T]]:
    """Mark a type as extending a type from another subgraph.

    Extended types can have external fields from the authoritative subgraph
    and add new fields specific to this subgraph.

    Args:
        cls: The class being decorated.

    Returns:
        The original class with federation.extend = True.

    Example:
        ```python
        @fraiseql_type
        @extends
        @key("id")
        class User:
            id: str = external()
            email: str = external()
            orders: list[Order]  # New field in this subgraph
        ```

    Raises:
        ValueError: If @extends is used without @key decorator.
    """

    def decorator(c: type[T]) -> type[T]:
        # Get or initialize federation metadata
        metadata = _get_or_init_federation_metadata(c)

        # Validate @key decorator was applied first
        if not metadata[_KEYS]:
            raise FederationValidationError(f"@extends requires @key decorator on {c.__name__}")

        # Mark type as extended
        metadata[_EXTEND] = True

        # Extract key field names for validation
        key_fields = _extract_key_fields(metadata)

        # Collect field markers and check consistency
        annotations = getattr(c, "__annotations__", {})
        _, field_markers = _collect_field_markers(c)

        # Determine which non-key fields are external
        external_non_key = {
            name
            for name, marker in field_markers.items()
            if marker.external and name not in key_fields
        }

        # If any non-key field is external, all key fields must be external too
        if external_non_key:
            external_key_fields = {
                name
                for name, marker in field_markers.items()
                if marker.external and name in key_fields
            }
            missing_key_fields = key_fields - external_key_fields
            if missing_key_fields:
                # Report error using first non-key external field (for consistent error)
                raise FederationValidationError(
                    f"Field '{next(iter(external_non_key))}' not found in {c.__name__}"
                )

        # Process field markers (requires, provides, external)
        _process_field_markers(metadata, field_markers, annotations, c.__name__, is_extended=True)

        return c

    # Support both @extends and @extends()
    if cls is None:
        # Called with parentheses: @extends()
        return decorator
    # Called without parentheses: @extends
    return decorator(cls)

File: fraiseql-typescript/src/federation.ts
/**
 * Federation support for Apollo Federation v2 in FraiseQL.
 *
 * This module provides decorators for defining federated GraphQL schemas:
 * - @Key: Define federation keys for entity resolution
 * - @Extends: Extend types from other subgraphs
 * - @External: Mark fields as owned by other subgraphs
 * - @Requires: Declare field dependencies
 * - @Provides: Mark fields that provide data for other subgraphs
 *
 * @example
 * ```typescript
 * import { Type, Key, Extends, External, Requires, Provides } from "fraiseql";
 *
 * // Authoritative User type in this subgraph
 * @Key("id")
 * @Type()
 * class User {
 *   id: string;
 *   email: string;
 * }
 *
 * // Extended User type in another subgraph
 * @Extends()
 * @Key("id")
 * @Type()
 * class User {
 *   @External() id: string;
 *   @External() email: string;
 *   orders: Order[];  // New field in this subgraph
 * }
 * ```
 */

import { SchemaRegistry } from "./registry";

/**
 * Global metadata storage for federation field markers.
 */
const fieldMetadata = new Map<object, Map<string, FieldMarker>>();

/**
 * Track field names per class for validation.
 */
const classFields = new Map<any, Set<string>>();

export type { ID } from "./scalars";

/**
 * Augment class constructor to support __fraiseqlFederation__ property.
 */
declare global {
  interface Function {
    __fraiseqlFederation__: FederationMetadata;
    __fraiseqlType__: boolean;
  }
}

/**
 * Federation metadata stored on type classes.
 */
interface FederationMetadata {
  keys: Array<{ fields: string[] }>;
  extend: boolean;
  external_fields: string[];
  requires: Record<string, string>;
  provides_data: string[];
}

/**
 * Property descriptor for federation field markers.
 */
class FieldMarker {
  constructor(
    public external: boolean = false,
    public requires: string | null = null,
    public provides: string[] = []
  ) {}
}

/**
 * Mark a field as external (owned by another subgraph).
 *
 * Use this in extended types to mark which fields are owned by the
 * authoritative subgraph.
 *
 * @example
 * ```typescript
 * @Extends()
 * @Key("id")
 * @Type()
 * class User {
 *   @External() id: string;
 *   @External() email: string;
 *   orders: Order[];  // New field in this subgraph
 * }
 * ```
 */
export function External(): PropertyDecorator {
  return function (target: object, propertyKey: string | symbol | undefined) {
    if (!propertyKey) return;
    const key = String(propertyKey);
    if (!fieldMetadata.has(target)) {
      fieldMetadata.set(target, new Map());
    }
    fieldMetadata.get(target)!.set(key, new FieldMarker(true));
  };
}

/**
 * Mark a field as requiring another field to be resolved first.
 *
 * This declares that a field needs data from another field (in the same
 * type or from federation) to compute its value.
 *
 * @param fieldName Name of the field that must be resolved first
 *
 * @example
 * ```typescript
 * @Extends()
 * @Key("id")
 * @Type()
 * class User {
 *   @External() id: string;
 *   @External() email: string;
 *   @Requires("email")
 *   profile: UserProfile;  // Needs email to resolve
 * }
 * ```
 */
export function Requires(fieldName: string): PropertyDecorator {
  return function (target: object, propertyKey: string | symbol | undefined) {
    if (!propertyKey) return;
    const key = String(propertyKey);

    // Find the class from the prototype
    const classConstructor = (target as any).constructor;
    const fieldSet = classFields.get(classConstructor);

    // Validate required field exists
    if (fieldSet && !fieldSet.has(fieldName)) {
      throw new Error(`Field '${fieldName}' not found`);
    }

    if (!fieldMetadata.has(target)) {
      fieldMetadata.set(target, new Map());
    }
    fieldMetadata.get(target)!.set(key, new FieldMarker(false, fieldName));
  };
}

/**
 * Mark a field as providing data for external subgraph fields.
 *
 * This declares that this field's data can be used to resolve fields
 * in other subgraphs.
 *
 * @param targets List of "Type.field" references this field provides data for
 *
 * @example
 * ```typescript
 * @Key("id")
 * @Type()
 * class User {
 *   id: string;
 *   @Provides("Order.owner_email", "Invoice.owner_email")
 *   email: string;
 * }
 * ```
 */
export function Provides(...targets: string[]): PropertyDecorator {
  return function (target: object, propertyKey: string | symbol | undefined) {
    if (!propertyKey) return;
    const key = String(propertyKey);
    if (!fieldMetadata.has(target)) {
      fieldMetadata.set(target, new Map());
    }
    fieldMetadata.get(target)!.set(key, new FieldMarker(false, null, targets));
  };
}

/**
 * Mark a type with a federation key for entity resolution.
 *
 * Federation keys are used to uniquely identify entities and resolve them
 * across subgraphs. Multiple @Key decorators define composite keys.
 *
 * @param fieldNames Field name or list of field names that form the key
 *
 * @example
 * ```typescript
 * @Key("id")
 * @Type()
 * class User {
 *   id: string;
 *   email: string;
 * }
 *
 * @Key("tenant_id")
 * @Key("id")
 * @Type()
 * class Account {
 *   tenant_id: string;
 *   id: string;
 *   name: string;
 * }
 * ```
 */
export function Key(
  fieldNames: string | string[]
): ClassDecorator & PropertyDecorator {
  // When used as class decorator (called first)
  return function (target: any) {
    const fields =
      typeof fieldNames === "string" ? [fieldNames] : fieldNames;

    // Initialize federation metadata if not present
    if (!target.__fraiseqlFederation__) {
      target.__fraiseqlFederation__ = {
        keys: [],
        extend: false,
        external_fields: [],
        requires: {},
        provides_data: [],
      };
    }

    const metadata: FederationMetadata = target.__fraiseqlFederation__;

    // Validate that @Type decorator was applied
    if (!target.__fraiseqlType__) {
      throw new TypeError(`@Key requires @Type decorator to be applied to ${target.name}`);
    }

    // Validate that key fields exist in the class
    const classFieldSet = classFields.get(target);
    if (classFieldSet) {
      for (const fieldName of fields) {
        if (!classFieldSet.has(fieldName)) {
          throw new Error(`Field '${fieldName}' not found`);
        }
      }
    }

    // Check for duplicate keys
    const newKey = { fields };
    const isDuplicate = metadata.keys.some(
      (existingKey) =>
        JSON.stringify(existingKey.fields) === JSON.stringify(newKey.fields)
    );

    if (isDuplicate) {
      throw new Error(`Duplicate key field in ${target.name}`);
    }

    // Add key to federation metadata
    metadata.keys.push(newKey);

    return target;
  } as any;
}

/**
 * Mark a type as extending a type from another subgraph.
 *
 * Extended types can have external fields from the authoritative subgraph
 * and add new fields specific to this subgraph.
 *
 * @example
 * ```typescript
 * @Extends()
 * @Key("id")
 * @Type()
 * class User {
 *   @External() id: string;
 *   @External() email: string;
 *   orders: Order[];  // New field in this subgraph
 * }
 * ```
 */
export function Extends(): ClassDecorator {
  return function (target: any) {
    // Initialize federation metadata if not present
    if (!target.__fraiseqlFederation__) {
      target.__fraiseqlFederation__ = {
        keys: [],
        extend: false,
        external_fields: [],
        requires: {},
        provides_data: [],
      };
    }

    const metadata: FederationMetadata = target.__fraiseqlFederation__;

    // Check that @Key decorator was used
    if (!metadata.keys || metadata.keys.length === 0) {
      throw new Error(`@Extends requires @Key decorator on ${target.name}`);
    }

    // Mark type as extended
    metadata.extend = true;

    // Extract key field names
    const keyFields = new Set<string>();
    for (const keyDef of metadata.keys) {
      for (const field of keyDef.fields) {
        keyFields.add(field);
      }
    }

    // Scan fields for field markers
    const prototype = target.prototype || target;
    const externalFields = new Set<string>();

    if (fieldMetadata.has(prototype)) {
      const classMetadata = fieldMetadata.get(prototype)!;
      for (const [key, marker] of classMetadata.entries()) {
        if (marker instanceof FieldMarker) {
          if (marker.external) {
            externalFields.add(key);
          }
          if (marker.requires) {
            metadata.requires[key] = marker.requires;
          }
          if (marker.provides && marker.provides.length > 0) {
            metadata.provides_data.push(...marker.provides);
          }
        }
      }
    }

    // Store external fields
    metadata.external_fields = Array.from(externalFields);

    return target;
  };
}

/**
 * Mark a class as a GraphQL type with optional federation support.
 *
 * @example
 * ```typescript
 * @Type()
 * class User {
 *   id!: string;
 *   email!: string;
 * }
 * ```
 */
export function Type(): ClassDecorator {
  return function (target: any) {
    // Mark that @Type decorator was applied
    target.__fraiseqlType__ = true;

    // Initialize federation metadata if not present
    if (!target.__fraiseqlFederation__) {
      target.__fraiseqlFederation__ = {
        keys: [],
        extend: false,
        external_fields: [],
        requires: {},
        provides_data: [],
      };
    }

    // Collect field names from class definition for validation
    const fieldSet = new Set<string>();
    const prototype = target.prototype || target;

    // Get field names from property descriptors
    for (const key of Object.getOwnPropertyNames(prototype)) {
      if (key !== 'constructor') {
        fieldSet.add(key);
      }
    }

    // Also collect from class instance if in test environment
    const instance = new target();
    for (const key in instance) {
      if (instance.hasOwnProperty(key) || Object.getPrototypeOf(instance).hasOwnProperty(key)) {
        fieldSet.add(key);
      }
    }

    // Store field names for this class
    classFields.set(target, fieldSet);

    const metadata: FederationMetadata = target.__fraiseqlFederation__;

    // Validate that @external is only used on extended types
    if (fieldMetadata.has(prototype)) {
      const classMetadata = fieldMetadata.get(prototype)!;
      for (const [fieldName, marker] of classMetadata.entries()) {
        if (marker instanceof FieldMarker) {
          // Validate field exists
          if (!fieldSet.has(fieldName)) {
            throw new Error(`Field '${fieldName}' not found`);
          }

          if (marker.external) {
            if (!metadata.extend) {
              throw new Error("@external requires @extends");
            }
          }

          if (marker.requires) {
            // Validate required field exists
            if (!fieldSet.has(marker.requires)) {
              throw new Error(`Field '${marker.requires}' not found`);
            }
          }
        }
      }
    }

    // Register type with schema registry
    const fields: Record<string, any> = {};
    for (const fieldName of fieldSet) {
      fields[fieldName] = {
        type: "String", // Default type for property markers
        nullable: false,
      };
    }

    // Register with schema registry (federation metadata will be added in schema generation)
    try {
      const fieldArray: Array<{
        name: string;
        type: string;
        nullable: boolean;
      }> = Object.entries(fields).map(([name, info]) => ({
        name,
        ...info,
      }));
      SchemaRegistry.registerType(target.name, fieldArray, undefined);
    } catch (e) {
      // Ignore registry errors in tests
    }

    return target;
  };
}

/**
 * Generate schema JSON from the current registry.
 *
 * @param types Optional list of types to include
 * @returns Schema dictionary with federation metadata if applicable
 */
export function generateSchemaJson(
  types?: Array<{ name: string }> | unknown
): Record<string, any> {
  const schema = SchemaRegistry.getSchema();
  let hasFederation = false;

  // Enhance types with federation metadata
  const enhancedTypes: Record<string, any>[] = [];

  if (schema.types && Array.isArray(schema.types)) {
    for (const typeInfo of schema.types) {
      // Check if any type with this name has federation metadata
      const typeClass = Array.isArray(types)
        ? types.find((t: any) => t.name === typeInfo.name)
        : null;

      const metadata = typeClass?.__fraiseqlFederation__;

      if (metadata) {
        hasFederation = true;
        const enhancedType: any = { ...typeInfo };

        // Add type-level federation metadata
        enhancedType.federation = {
          keys: metadata.keys,
          extend: metadata.extend,
          external_fields: metadata.external_fields,
          shareable: false,
        };

        // Add field-level federation metadata
        if (enhancedType.fields && Array.isArray(enhancedType.fields)) {
          enhancedType.fields = enhancedType.fields.map((field: any) => {
            const fieldMarker = fieldMetadata.get(typeClass?.prototype || typeClass);
            const marker = fieldMarker?.get(field.name);

            const fieldFederation: Record<string, any> = {};
            let hasfieldFederation = false;

            if (marker instanceof FieldMarker) {
              if (marker.external) {
                fieldFederation.external = true;
                hasfieldFederation = true;
              }
              if (marker.requires) {
                fieldFederation.requires = marker.requires;
                hasfieldFederation = true;
              }
              if (marker.provides && marker.provides.length > 0) {
                fieldFederation.provides = marker.provides;
                hasfieldFederation = true;
              }
            }

            return hasfieldFederation ? { ...field, federation: fieldFederation } : field;
          });
        }

        enhancedTypes.push(enhancedType);
      } else {
        enhancedTypes.push(typeInfo);
      }
    }
  }

  return {
    ...schema,
    types: enhancedTypes,
    federation: {
      enabled: hasFederation,
      version: "v2",
    },
  };
}

// Re-export commonly used types
export { SchemaRegistry };
